### PR TITLE
Hashmap ensure replace

### PR DESCRIPTION
**Pull request by teknoraver**
Add a helper to replace hashmap members while ensuring it's allocated.

This helper is pulled from PR #32212

### CODE REVIEW

File: src/test/test-hashmap-plain.c [hunk 1]
```diff
@@ -49,6 +49,35 @@ TEST(hashmap_replace) {
         ASSERT_STREQ(r, "val5");
 }
 
+TEST(hashmap_ensure_replace) {
+        _cleanup_hashmap_free_ Hashmap *m = NULL;
+        _cleanup_free_ char *val1 = NULL, *val2 = NULL;
+        char *s;
```
Comment (daandemeyer) [inline src/test/test-hashmap-plain.c]: I would get rid of this variable and just inline every call to hashmap_get() into the assertion

File: src/test/test-hashmap-plain.c [hunk 2]
```diff
@@ -49,6 +49,35 @@ TEST(hashmap_replace) {
...
+
+        val1 = strdup("val1");
+        assert_se(val1);
```
Comment (daandemeyer) [inline src/test/test-hashmap-plain.c]: ASSERT_NOT_NULL

File: src/test/test-hashmap-plain.c [hunk 3]
```diff
@@ -49,6 +49,35 @@ TEST(hashmap_replace) {
...
+        val2 = strdup("val2");
+        assert_se(val2);
```
Comment (daandemeyer) [inline src/test/test-hashmap-plain.c]: ditto

File: src/test/test-hashmap-plain.c [hunk 4]
```diff
@@ -49,6 +49,35 @@ TEST(hashmap_replace) {
...
+
+        assert_se(hashmap_ensure_replace(&m, &string_hash_ops, val1, val2) >= 0);
```
Comment (daandemeyer) [inline src/test/test-hashmap-plain.c]: ASSERT_OK

File: src/test/test-hashmap-plain.c [hunk 5]
```diff
@@ -49,6 +49,35 @@ TEST(hashmap_replace) {
...
+
+        assert_se(hashmap_ensure_replace(&m, &string_hash_ops, "key 1", val1) >= 0);
```
Comment (daandemeyer) [inline src/test/test-hashmap-plain.c]: ditto here and below

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

**Review by DaanDeMeyer — APPROVED** [review]

File: src/libsystemd/sd-hwdb/sd-hwdb.c [hunk 6]
```diff
@@ -166,11 +166,7 @@ static int hwdb_add_property(sd_hwdb *hwdb, const struct trie_value_entry_f *ent
                 }
         }
 
-        r = ordered_hashmap_ensure_allocated(&hwdb->properties, &string_hash_ops);
-        if (r < 0)
-                return r;
-
-        r = ordered_hashmap_replace(hwdb->properties, key, (void *)entry);
+        r = ordered_hashmap_ensure_replace(&hwdb->properties, &string_hash_ops, key, (void *)entry);
```
Comment (YHNdnzj) [inline src/libsystemd/sd-hwdb/sd-hwdb.c]: If you touch this:


```suggestion
        r = ordered_hashmap_ensure_replace(&hwdb->properties, &string_hash_ops, key, (void *) entry);
```

**Review by YHNdnzj — APPROVED** [review]

**Review by bluca — APPROVED** [review]

**Pull request merged**